### PR TITLE
fix sanity test DeleteVolume failure

### LINKED ISSUE #27: fix sanity test failures
**What happened**:

```
Summarizing 5 Failures:

[Fail] Controller Service CreateVolume [It] should not fail when requesting to create a volume with already existing name and same capacity. 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:371

[Fail] Controller Service CreateVolume [It] should fail when requesting to create a volume with already existing name and different capacity. 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:442

[Fail] Controller Service ValidateVolumeCapabilities [It] should fail when the requested volume does not exist 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:705

[Fail] Node Service [It] should work 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/node.go:464

Ran 26 of 57 Specs in 77.639 seconds
FAIL! -- 21 Passed | 5 Failed | 0 Pending | 31 Skipped
```

following issue has been fixed:
```
[Fail] Controller Service DeleteVolume [It] should succeed when an invalid volume id is used 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:543
```
**What you expected to happen**:

**How to reproduce it**:
run sanity test: https://github.com/kubernetes-sigs/azurefile-csi-driver/tree/master/test/sanity

**Anything else we need to know?**:

**Environment**:
- CSI Driver version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by lizebang**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #27 

**Special notes for your reviewer**:

**Release note**:

```
fix sanity test DeleteVolume failure
```

### CODE REVIEW

File: pkg/azurefile/controllerserver.go [hunk 1]
```diff
@@ -19,11 +19,13 @@ package azurefile
 import (
 	"context"
 	"fmt"
+	"net/http"
 	"strings"
 
 	"k8s.io/kubernetes/pkg/volume/util"
 
 	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2018-07-01/storage"
+	check "github.com/Azure/azure-sdk-for-go/storage"
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: `check` does not make sense, pls remove it

File: pkg/azurefile/controllerserver.go [hunk 2]
```diff
@@ -112,18 +114,21 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
...
 
 	if err := d.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME); err != nil {
 		return nil, fmt.Errorf("invalid delete volume req: %v", req)
 	}
 
-	volumeID := req.VolumeId
 	resourceGroupName, accountName, fileShareName, err := getFileShareInfo(volumeID)
 	if err != nil {
-		return nil, err
+		// This operation MUST be idempotent. If a volume corresponding to the specified volume_id does not
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: remove the comments

File: pkg/azurefile/controllerserver.go [hunk 3]
```diff
@@ -112,18 +114,21 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
...
+		// exist or the artifacts associated with the volume do not exist anymore, the Plugin MUST reply 0 OK.
+		// reference: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume
+		return &csi.DeleteVolumeResponse{}, nil
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: `return &csi.DeleteVolumeResponse{}, err`

File: pkg/azurefile/controllerserver.go [hunk 4]
```diff
@@ -132,11 +137,20 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
 
 	accountKey, err := GetStorageAccesskey(d.cloud, accountName, resourceGroupName)
 	if err != nil {
-		return nil, fmt.Errorf("no key for storage account(%s) under resource group(%s), err %v", accountName, resourceGroupName, err)
+		// Same case as above
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: remove comments

File: pkg/azurefile/controllerserver.go [hunk 5]
```diff
@@ -132,11 +137,20 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
...
+		// reference: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume
+		return &csi.DeleteVolumeResponse{}, nil
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: `return &csi.DeleteVolumeResponse{}, err`

File: pkg/azurefile/controllerserver.go [hunk 6]
```diff
@@ -132,11 +137,20 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
...
 	klog.V(2).Infof("deleting azure file(%s) rg(%s) account(%s) volumeID(%s)", fileShareName, resourceGroupName, accountName, volumeID)
 	if err := d.cloud.DeleteFileShare(accountName, accountKey, fileShareName); err != nil {
+		if e, ok := err.(check.AzureStorageServiceError); ok && e.StatusCode == http.StatusNotFound {
+			// Same case as above
+			// reference: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume
+			return &csi.DeleteVolumeResponse{}, nil
+		}
+		// TODO: add ShareHasSnapshots check
+		// reference: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume
 		return nil, err
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: may be `return &csi.DeleteVolumeResponse{}, err`?

File: pkg/azurefile/controllerserver.go [hunk 7]
```diff
@@ -132,11 +137,20 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
...
 	if err != nil {
-		return nil, fmt.Errorf("no key for storage account(%s) under resource group(%s), err %v", accountName, resourceGroupName, err)
+		// Same case as above
+		// reference: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume
+		return &csi.DeleteVolumeResponse{}, nil
 	}
 
 	klog.V(2).Infof("deleting azure file(%s) rg(%s) account(%s) volumeID(%s)", fileShareName, resourceGroupName, accountName, volumeID)
 	if err := d.cloud.DeleteFileShare(accountName, accountKey, fileShareName); err != nil {
+		if e, ok := err.(check.AzureStorageServiceError); ok && e.StatusCode == http.StatusNotFound {
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: the status check may not make sense

**Comment by andyzhangx** [comment]
BTW, @lizebang you could refer to azure disk driver implementation:
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/2ceef88dea97986e9c1c01348a8347a82689372f/pkg/azuredisk/controllerserver.go#L240-L263

**Comment by lizebang** [comment]
@andyzhangx Yeah, I'll do it now.

**Comment by lizebang** [comment]
@andyzhangx 
I modified the code and tested it, and it failed.

By referring to the CSI spec and csi-test lib and some other csi-driver implementations, it should `return &csi.DeleteVolumeResponse{}, nil` when the volume ID does not exist.

Reference:
csi-test lib:
https://github.com/kubernetes-csi/csi-test/blob/5b1e3786b7c8f7ca514b40e882a0b5dc36e4c842/pkg/sanity/controller.go#L543

some other csi-driver implementations
gcp-compute-persistent-disk-csi-driver:
https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/5d62c770f3e331e4236ebe42b74321f007f8c1e7/pkg/gce-pd-csi-driver/controller.go#L211
gcp-filestore-csi-driver:
https://github.com/kubernetes-sigs/gcp-filestore-csi-driver/blob/e961734c5b26ed5a9634a20f59fc476b3a8433b2/pkg/csi_driver/controller.go#L172
aws-ebs-csi-driver:
https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/7632fba99f7efb28d21ffb3d451c0800cd3ca2b5/pkg/driver/controller.go#L158
aws-fsx-csi-driver:
https://github.com/kubernetes-sigs/aws-fsx-csi-driver/blob/761ce9451ba4be86a0c6c457171f0369326d34a1/pkg/driver/controller.go#L107

File: pkg/azurefile/controllerserver.go [hunk 8]
```diff
@@ -132,11 +137,20 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
...
+			// Same case as above
+			// reference: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume
+			return &csi.DeleteVolumeResponse{}, nil
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: if delete volume failed, it should not `return &csi.DeleteVolumeResponse{}, nil`, right?

**Comment by andyzhangx** [comment]
> By referring to the CSI spec and csi-test lib and some other csi-driver implementations, it should return &csi.DeleteVolumeResponse{}, nil when the volume ID does not exist.

Correct, while if delete volume failed, it should return error. and pls remove the comments, thanks.

**Comment by andyzhangx** [comment]
BTW, the test failed by:
```
Verifying dep check
# Gopkg.lock is out of sync:
github.com/Azure/azure-sdk-for-go/storage: imported or required, but missing from Gopkg.lock's input-imports
The command "hack/verify-all.sh" exited with 1.
```

Could you run `hack/verify-all.sh` on your side to check?

**Comment by andyzhangx** [comment]
BTW, for the non exist volume, pls log the error, otherwise there is error message missing

**Comment by lizebang** [comment]
@andyzhangx Thanks, I have removed the comments and will check more.

File: pkg/azurefile/controllerserver.go [hunk 9]
```diff
@@ -117,13 +117,14 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
...
 	if err := d.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME); err != nil {
-		return nil, fmt.Errorf("invalid delete volume req: %v", req)
+		return nil, status.Errorf(codes.InvalidArgument, "invalid delete volume request: %v", req)
 	}
 
 	volumeID := req.VolumeId
 	resourceGroupName, accountName, fileShareName, err := getFileShareInfo(volumeID)
 	if err != nil {
-		return nil, err
+		klog.V(2).Infof("azure file(%s) under rg(%s) account(%s) volumeID(%s) is invalid, returning with success: %v", fileShareName, resourceGroupName, accountName, volumeID, err)
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: `klog.Errorf("getFileShareInfo(%s) in DeleteVolume failed with error: %v`, volumeID)`

File: pkg/azurefile/controllerserver.go [hunk 10]
```diff
@@ -132,12 +133,12 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
 
 	accountKey, err := GetStorageAccesskey(d.cloud, accountName, resourceGroupName)
 	if err != nil {
-		return nil, fmt.Errorf("no key for storage account(%s) under resource group(%s), err %v", accountName, resourceGroupName, err)
+		return nil, status.Errorf(codes.Internal, "could not get the storage access key: %v", err)
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: would you keep the original code:
`return nil, fmt.Errorf("no key for storage account(%s) under resource group(%s), err %v", accountName, resourceGroupName, err)`

File: pkg/azurefile/controllerserver.go [hunk 11]
```diff
@@ -132,12 +133,12 @@ func (d *Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest)
...
 	}
 
 	klog.V(2).Infof("deleting azure file(%s) rg(%s) account(%s) volumeID(%s)", fileShareName, resourceGroupName, accountName, volumeID)
 	if err := d.cloud.DeleteFileShare(accountName, accountKey, fileShareName); err != nil {
-		return nil, err
+		return nil, status.Errorf(codes.Internal, "could not delete the volume: %v", err)
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: `return nil, status.Errorf(codes.Internal, fmt.Errorf("DeleteFileShare %s under %s failed with error: %v", fileShareName, accountName, err)`

Comment (lizebang) [reply pkg/azurefile/controllerserver.go]: I think here is `return nil, status.Errorf(codes.Internal, "DeleteFileShare %s under %s failed with error: %v", fileShareName, accountName, err)`.

**Comment by andyzhangx** [comment]
Hi @lizebang, I have changed your PR title to make it more clear and also added release-note.
Pls do it when you send PR next time, thanks.

**Comment by lizebang** [comment]
@andyzhangx 
Thanks a lot, I'll pay attention to this.
And pls reopen the issue #27 , the k8s-ci-robot mistakenly closed it.

**Pull request merged**